File: tests/toolchain/variadic_probe.zig
const std = @import("std");

pub export fn sqlite_zig_variadic_dispatch(
    op: c_int,
    int_arg: c_int,
    ptr_arg: ?*anyopaque,
    i64_arg: i64,
    double_arg: f64,
) callconv(.c) c_int {
    return switch (op) {
        0 => 100,
        1 => if (int_arg == 42) 142 else -101,
        2 => if (ptr_arg != null and int_arg == 7) 207 else -102,
        3 => if (i64_arg == 0x102030405060708 and double_arg == 2.5) 325 else -103,
        else => -104,
    };
}

extern fn run_variadic_probe() callconv(.c) c_int;

pub fn main() !void {
    const rc = run_variadic_probe();
    if (rc != 0) {
        std.log.err("variadic ABI probe failed at case {d}", .{rc});
        return error.VariadicAbiProbeFailed;
    }
}

File: tests/toolchain/variadic_probe.c
#include <stdarg.h>
#include <stdint.h>

extern int sqlite_zig_variadic_dispatch(
    int op,
    int int_arg,
    void *ptr_arg,
    int64_t i64_arg,
    double double_arg);

static int probe_config_shape(int op, ...) {
  int int_arg = 0;
  void *ptr_arg = 0;
  int64_t i64_arg = 0;
  double double_arg = 0.0;
  va_list args;
  va_start(args, op);
  switch (op) {
    case 0:
      break;
    case 1:
      int_arg = va_arg(args, int);
      break;
    case 2:
      ptr_arg = va_arg(args, void *);
      int_arg = va_arg(args, int);
      break;
    case 3:
      i64_arg = va_arg(args, int64_t);
      double_arg = va_arg(args, double);
      break;
    default:
      va_end(args);
      return -1;
  }
  va_end(args);
  return sqlite_zig_variadic_dispatch(
      op, int_arg, ptr_arg, i64_arg, double_arg);
}

int run_variadic_probe(void) {
  int marker = 9;
  if (probe_config_shape(0) != 100) return 1;
  if (probe_config_shape(1, 42) != 142) return 2;
  if (probe_config_shape(2, &marker, 7) != 207) return 3;
  if (probe_config_shape(3, INT64_C(0x102030405060708), 2.5) != 325) return 4;
  return 0;
}
